### ISSUE TITLE
React Hook useEffect has a missing dependency: 'xxx' ?

### ISSUE THREAD

**Issue by luojinghui** [facebook/react#15865]
Hi, I am researching React Hook recently, it's great.

But I have a problem, I did not find a suitable answer in the React Hook documentation and google to solve my problem.

When I used a function in useEffect and function component together, I encountered a warning that I did not specify a dependency in useEffect, as shown below:

![image](https://user-images.githubusercontent.com/12367406/59340801-4910c980-8d39-11e9-9c6d-41a7b95e06cb.png)

Is there any way to solve my problem? I have thought about it for a long time.

**What is the current behavior?**
```
React Hook useEffect has a missing dependency: 'setCenterPosition'. Either include it or remove the dependency array. (react-hooks/exhaustive-deps)
```

**Mini repro in codesandbox:**
https://codesandbox.io/s/trusting-kowalevski-oet4b

Any solution, thanks.

**Code**
```jsx
function App() {
  const [elePositionArr, setElePositionArr] = useState([
    { left: 0, top: 0 },
    { left: 0, top: 0 },
    { left: 0, top: 0 }
  ]);
  const stageRef = useRef(null);
  const Stage = useRef({ w: 0, h: 0 });
  const currentIndex = useRef(0);

  useEffect(() => {
    // Record the size of the stage
    const stageW = stageRef.current.scrollWidth;
    const stageH = stageRef.current.scrollHeight;

    Stage.current = { w: stageW, h: stageH };

    const index = Math.floor(Math.random() * 3);
    currentIndex.current = index;
    setCenterPosition(index);
  }, []);

  // Centering a block element
  function setCenterPosition(index) {
    let cacheEle = elePositionArr;
    // calc center postion
    const centerOfLeft = Stage.current.w / 2 - 25;
    const centerOfTop = Stage.current.h / 2 - 25;

    cacheEle = cacheEle.map((item, i) => {
      const randomWNum = Math.floor(Math.random() * Stage.current.w) - 50;
      const randomHNum = Math.floor(Math.random() * Stage.current.h) - 50;
      const randomLeft = randomWNum <= 50 ? 50 : randomWNum;
      const randomTop = randomHNum <= 50 ? 50 : randomHNum;
      let newItem;

      if (index === i) {
        newItem = { left: centerOfLeft, top: centerOfTop };
      } else {
        newItem = { left: randomLeft, top: randomTop };
      }

      return newItem;
    });

    setElePositionArr(cacheEle);
  }

  function handleClickLi(index) {
    if (currentIndex.current !== index) {
      setCenterPosition(index);
      currentIndex.current = index;
    }
  }

  return (
    <div className="container">
      <div className="stage" ref={stageRef}>
        {elePositionArr.map((item, index) => (
          <div className="ele" key={index} style={item}>
            {index}
          </div>
        ))}
      </div>
      <ul className="nav">
        {elePositionArr.map((item, index) => (
          <li
            className={currentIndex.current === index ? "active-li" : ""}
            onClick={() => {
              handleClickLi(index);
            }}
            key={"li" + index}
          >
            {index}
          </li>
        ))}
      </ul>
    </div>
  );
}
```

**Comment by gaearon** [comment]
This should answer your question:

https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies

**Comment by luojinghui** [comment]
@gaearon 
Did not solve my problem, my problem is more special, my run function`setElePositionArr` wants to fire on componentDidMount and handleClick, but the React Hook function does not meet my requirements and a warning appears.

**Comment by luojinghui** [comment]
> This should answer your question:
> 
> https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies

Did not solve my problem, my problem is more special, my run functionsetElePositionArr wants to fire on componentDidMount and handleClick, but the React Hook function does not meet my requirements and a warning appears.

**Comment by luojinghui** [comment]
My question is :React Hook cannot call a function in the useEffect and function component to setState operation？

**Comment by gaearon** [comment]
Leaving open so we can respond to second question.

**Comment by asumaran** [comment]
I'm having the same ESLint warning but in my case I want to make an async call when my component is mounted.

```tsx
const ManageTagsModalBody: React.FC<IProps> = ({ children, getTags }) => {
  useEffect(() => {
    getTags();
  }, []);

  return <div>{children}</div>;
};
```


```
Line 12:  React Hook useEffect has a missing dependency: 'getTags'. Either include it or remove the dependency array. If 'getTags' changes too often, find the parent component that defines it and wrap that definition in useCallback  react-hooks/exhaustive-deps 
```

`ManageTagsModalBody` doesn't depend on any data it's only used as a wrapper to load data anytime the component is rendered. But it's worth noting the `children` component is using the data `getTags` has fetch. Not sure what to add to the list of dependencies.

**Comment by MS44118** [comment]
i'm interested for an explanation also.
 **React Hook useEffect has a missing dependency: 'dispatch'.**

```
  useEffect(() => {
    axios.get('http://localhost:8000/api/future-registrations')
      .then((result) => {
        dispatch(initRegistrationsAction(result.data));
      });
  }, []);
```

**Comment by asumaran** [comment]
@MS44118 that’s because `dispatch` is not available in your scope. you need to create a prop like `initRegistrations` which contains your `dispatch(initRegistrationsAction(result.data))` in your `mapDispatchToProps` function and use that instead of dispatching the action directly in the component.

**Comment by ifokeev** [comment]
Same issue here. Strange rule.

**Comment by OnkelTem** [comment]
Same here. As simple as in https://github.com/facebook/react/issues/15865#issuecomment-506503377:

```
  useEffect(() => {
    dispatch(fetchPosts());
  }, []);
```

Is there a way to suppress this warning?

**Comment by fayway** [comment]
As suggested in the warning message, you can do like that

```
const initFetch = useCallback(() => {
    dispatch(fetchPosts());
  }, [dispatch]);

  useEffect(() => {
    initFetch();
  }, [initFetch]);
```

**Comment by shkyung** [comment]
then whenver initFetch and dispatch value changed, useEffect's callback will be execute.
he wants to execute callback once

**Comment by fayway** [comment]
@shkyung In my case, **initFetch** will not change cause (store.)**dispatch** doesn't either

**Comment by praganavictor** [comment]
I'm having the same ESLint warning and I trying to learn more about react hooks but I can't solved
React Hook useEffect has a missing dependency: 'dispatch'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

  useEffect(() => {
    async function buscaEventos() {
      const res = await api
        .get("/eventos/")
        .catch(err => console.error("err", err));
      dispatch(eventActions.SetEvents(res.data.eventos));
    }

    dispatch(navigationActions.setGoBack(false));
    buscaEventos();

    return () => dispatch(navigationActions.setGoBack(true));
  });

**Comment by loveonelong** [comment]
@fayway Maybe it's better
```jsx
const setCenterPosition = useRef(null)
setCenterPosition.current = useCallback( ()=>{} ,[deps])

effect(()=>{ setCenterPosition.current() },[setCenterPosition,otherDeps])
```

**Comment by wickstjo** [comment]
This change seems a bit ridiculous and counter productive to be quite frank. Rather rollback to an earlier version and let you think this one through.

**Comment by micahbales** [comment]
Have the same issue. I only want my fetch to fire on componentDidMount. Adding the dependency to the array results in repeatedly hitting the endpoint.

**Comment by Stoner609** [comment]
React Hook useEffect has a missing dependency: 'dispatch'.

```
const [state, dispatch] = useReducer(reducer, initialState);
const { count, step } = state;

useEffect(() => {
  const id = setInterval(() => {
    dispatch({ type: 'tick' }); // Instead of setCount(c => c + step);
  }, 1000);
  return () => clearInterval(id);
}, [dispatch]);
```

**Comment by rhulse** [comment]
This seems to be quite a common use case - one that I have run into on my current project. 

You want to run an effect hook just one time, but there IS a dependancy, although you only care about the state of that at startup. Currently using
 ```// eslint-disable-next-line react-hooks/exhaustive-deps```, 
 but would be nice to not have to in this case.

**Comment by luojinghui** [comment]
> This seems to be quite a common use case - one that I have run into on my current project.
> 
> You want to run an effect hook just one time, but there IS a dependancy, although you only care about the state of that at startup. Currently using
> `// eslint-disable-next-line react-hooks/exhaustive-deps`,
> to turn off the linting rule but would be nice to not have to in this case.

Closing Lint is just a temporary solution, the key to this problem is not enough to understand the hook.

**Comment by wot48** [comment]
Same issue here. I just wanted to fire a function inside useEffect() once, having [] as second parameter does what I want, but it keeps giving this warning.

**Comment by NordlingDev** [comment]
Such a weird warning because I've never seen it before and it ruins the use case of having something called only once when component has mounted. There doesn't seem to be any workaround except ignoring the lint warning.

**Comment by cynosurecreative** [comment]
Id like to put in my two cents here. In particular with using Axios to cancel a network request. Will create a separate issue if required but felt this issue directly affected myself. So my useApi hook is as follows -
```javascript
import axios, { AxiosInstance } from "axios";
import axiosRetry from "axios-retry";
import { FetchEnv } from "../../utilities";
import { useStorage } from "../useStorage";

export const useApi = ({ isAuth = false, retries = 3 } = {}) => {
  const cancelToken = axios.CancelToken.source();
  const { getItem } = useStorage();

  const getBaseUrl = () => {
    return FetchEnv({
      defaultValue: "http://api.example.com",
      key: "API_URI"
    });
  };

  const authorizedClient = (client: AxiosInstance) => {
    const session = getItem("SESSION", "sessionStorage");
    const hasAccessToken = Boolean(session.tokens.accessToken);

    if (isAuth && !hasAccessToken) {
      console.error("No access token found to initiate authorized request.");
      return client;
    } else {
      client.defaults.headers[
        "Authorization"
      ] = `Bearer ${session.tokens.accessToken}`;
      return client;
    }
  };

  const buildFetch = (): AxiosInstance => {
    const client = axios.create({
      baseURL: getBaseUrl(),
      cancelToken: cancelToken.token
    });
    axiosRetry(client, { retries });
    return isAuth ? authorizedClient(client) : client;
  };

  return {
    cancelRequest: cancelToken.cancel,
    fetch: buildFetch()
  };
};
```

To use this hook within a component, simply define it as below then it can be used to make both authenticated and unauthenticated calls either on mount, within a useEffect, as part of an event handler, etc.  The cancelRequest is the unique cancel token generated that is paired with the "fetch" axios instance relative to this component.

```javascript
const { cancelRequest, fetch } = useApi();
```

If for any reason this component dismounts, the following is called:

```javascript
  useEffect(() => {
    return () => {
      if (cancelRequest) {
        cancelRequest("Auth network request cancelled");
      }
    };
  }, []);
```

Now it works perfectly at present. However, the warning, when applied (putting cancelRequest as a useEffect dependency) immediately CANCELS the network request when the fetch method is called. Any advice would be greatly appreciated however it would appear the only solution at present would be to ts-lint disable this moving forward...

**Comment by rhulse** [comment]
I think the issue here **may** be the way we are thinking about hooks. @luojinghui's [comment](https://github.com/facebook/react/issues/15865#issuecomment-531704551) suggests that there is more to this than meets the eye. Perhaps we need to think differently? I have been able to remove this problem in one of my hooks with some refactoring. The code is clearer as a result. 

It led me to wonder if, in @Stoner609's [example](https://github.com/facebook/react/issues/15865#issuecomment-527297171) just above this, the timing code should be in a useCallback hook? It is still odd having a function (which should be static) as a dependancy though.

**Comment by Fieel** [comment]
I think this case should be some sort of exception

**Comment by jacquesadit** [comment]
If you want to run a function only once when the component loads and which takes in parameters then you can use useRef to avoid the warning. Something like this works:

```
  const InitialPropA= useRef(propA);
  useEffect(() => {
    myFunction(InitialPropA);
  }, [myFunction, InitialPropA]);
```

**Comment by Lukortech** [comment]
> As suggested in the warning message, you can do like that
> 
> ```
> const initFetch = useCallback(() => {
>     dispatch(fetchPosts());
>   }, [dispatch]);
> 
>   useEffect(() => {
>     initFetch();
>   }, [initFetch]);
> ```

As much as I love this witty reply this usually is what happens in my code. 
I don't want to have all those pesky warning about dispatch going missing. 

I'd love to know exactly what and why happens in that array of mystery :)

**Comment by GabeCabrera** [comment]
> As suggested in the warning message, you can do like that
> 
> ```
> const initFetch = useCallback(() => {
>     dispatch(fetchPosts());
>   }, [dispatch]);
> 
>   useEffect(() => {
>     initFetch();
>   }, [initFetch]);
> ```

this also may create a memory leak in case of something like an API, it will make the call on an infinite loop.

**Comment by Jordy-A** [comment]
Really need a fix for this. It keeps giving me errors and no ideas how to solve it if you want to fire a function inside `useEffect()` just once.

**Comment by vishwa3** [comment]
Facing the same error - 

**React Hook useEffect has a missing dependency: 'props'. Either include it or remove the dependency array. However, 'props' will change when *any* prop changes, so the preferred fix is to destructure the 'props' object outside of the useEffect call and refer to those specific props inside useEffect  react-hooks/exhaustive-deps**

Code - 

```
  useEffect(() => {
    props.dispatch(searchMediaAction("rain"));
  }, []);
```

**Comment by PatrickCavalcanti** [comment]
same error

**Comment by juanmartinzzz** [comment]
same problem. I want to imitate the behaviour of componentDidMount; I'm  not interested on the warning; I don't want to re-run the hook when anything changes

**Comment by asumaran** [comment]
@juanmartinzzz maybe this is what you need https://reactjs.org/docs/hooks-reference.html#conditionally-firing-an-effect

> If you want to run an effect and clean it up only once (on mount and unmount), you can pass an empty array ([]) as a second argument.

**Comment by tomwayson** [comment]
I would propose to **not** show this warning when the second argument to `useEffect()` is `[]`, b/c in that case the developer knows that this effect is only going to run once and therefore _wants_ to use the initial prop values and doesn't care if the prop changes on subsequent renders.

**Comment by prnsml** [comment]
Why is this closed?

**Comment by kennylbj** [comment]
Any solutions?

**Comment by robthedev** [comment]
@kennylbj 

After trying a bunch of different suggestions, this is what ultimately worked for me. I preferred not using eslint-disable, while trying to keep the code relatively simple for a newbie. This approach allows you to pass your dispatch function without invoking it inside of useCallback. Hope this helps.

```javascript
// ... inside function compoenent
const { dispatchFunc } = customHook();
const memoizeDispatchFunc = useCallback(dispatchFunc, []);

useEffect(() => {
  memoizeDispatchFunc();
}, [memoizeDispatchFunc]);
````

**Comment by oscadev** [comment]
Can we just remove the empty array? Doing so seems to work (code still runs as expected, and warning is gone), but I dont know if there is any reason we shouldn't do it that way.

`useEffect(()=>{
     myFunc();
     }
)`

**Comment by kennylbj** [comment]
@robthedev  Greate solution.
We can even simplify the code If we use redux:

```js
const dispatch = useDispatch();

// redux can guarantee that the dispatch function will not change between renders.
// so we don't need to wrap it with useCallback for now.
useEffect(() => {
  dispatch(actions());
}, [dispatch]);
```

Dan's [this article]( https://overreacted.io/a-complete-guide-to-useeffect) helps me a lot.

The solution to this problem is not to remove a dependency, instead, we can hoist functions that don’t need props or state outside of the component or, wrap them into useCallback where they’re defined.

**Comment by vouza** [comment]
> @kennylbj
> 
> After trying a bunch of different suggestions, this is what ultimately worked for me. I preferred not using eslint-disable, while trying to keep the code relatively simple for a newbie. This approach allows you to pass your dispatch function without invoking it inside of useCallback. Hope this helps.
> 
> ```js
> // ... inside function component
> const { dispatchFunc } = customHook();
> const memoizeDispatchFunc = useCallback(dispatchFunc, []);
> 
> useEffect(() => {
>   memoizeDispatchFunc();
> }, [memoizeDispatchFunc]);
> ```

With me the above example does not work.

Get this error:
`TypeError: Cannot destructure property 'dispatchFunc' of 'componentDidMount(...)' as it is undefined.`
Happy Boilerplate Code. There should be a nice

**Comment by rosanarezende** [comment]
For me, the error appeared in the following situation:
```
  useEffect(() => {
    props.getTasks()
  }, [])
```

I corrected it like this:
```
const { getTasks } = props
  useEffect(() => {
    getTasks()
  }, [getTasks])
```

**Comment by r3wt** [comment]
When thousands of devs face this same issue and the React dev team closes the thread and ignores everyone:
![8d6](https://user-images.githubusercontent.com/4483325/79016477-b2699080-7b34-11ea-807a-8adb73acad20.jpg)

**Comment by r3wt** [comment]
my issue was different than others in this thread, so i'll share it in case any other poor soul goes down the rabbit hole i just have been down for 1 1/2 hours
```jsx
function hydrate( _state, _errors=false) {
    console.log('hydrate()');
    setState({...state,..._state});
    if(_errors){
        setErrors({...errors,..._errors});
    }
}
```
the user is expected to use this with `useEffect()` in order to hydrate their state like so:

```jsx
useEffect(()=>{
    hydrate({
        name:'Garrett',
        email:'test@test.com',
        newsletter: 'yes'
    });
},[hydrate]); //unfortunately, this causes an infinite render, because hydrate will always change
```
so to solve this, i just changed my hydrate function to be wrapped in `useCallback()`, now it is not mutate every render ( i guess ), and supposedly this is better for performance, which if that turns to be true i will implement this for all of my helper function returned from my custom hook. 

```jsx
const hydrate = useCallback(( _state, _errors=false )=> {
    console.log('hydrate()');
    setState({...state,..._state});
    if(_errors){
        setErrors({...errors,..._errors});
    }
},[]);
```

can anyone confirm if this is true, useCallback is preventing `hydrate` from being mutated in the render, and thus it is probably better to wrap all such functions in useCallback for better performance?

**Comment by gugol2** [comment]
> ```js
> redux can guarantee that the dispatch function will not change between renders.
> ```

Are you sure about that? 
Link to the docs where it says so please.

**Comment by kennylbj** [comment]
Check https://overreacted.io/a-complete-guide-to-useeffect/#decoupling-updates-from-actions for more details or you may check [useEffect source code](https://github.com/facebook/react/blob/master/packages/react/src/ReactHooks.js) for implementation details.

Here I just reference what Dan said:

> The answer is that React guarantees the dispatch function to be constant throughout the component lifetime. So the example above doesn’t ever need to resubscribe the interval.

**Comment by gugol2** [comment]
> guarantees the dispatch function to be constant

Cool, thanks!

**Comment by aaj** [comment]
> When thousands of devs face this same issue and the React dev team closes the thread and ignores everyone:
> ![8d6](https://user-images.githubusercontent.com/4483325/79016477-b2699080-7b34-11ea-807a-8adb73acad20.jpg)

Pretty much this. Here I am giving hooks a try and I'm greeted with this warning which makes no sense and isn't covered in any of the official docs. Really makes you wonder if hooks are ready for serious work when a basic use case like "please just run this effect only once" gives this much trouble.

**Comment by edi** [comment]
Well, not much else to say, aside the fact that the thread is closed smh.

**Comment by scottwang329** [comment]
Faces the same issue!

**Comment by ra30r** [comment]
SOLUTION, 
(that worked for me):

```
const onInit = function(){ 
	console.log('initiated', Date.now());
}

useEffect(onInit, []);
```

I would prefer the issue to be fixed on linting side, but hey...

**Comment by Emidomenge** [comment]
@ra30r nice workaround but wondering if it can create side effects ? :thinking:

**Comment by ListeH** [comment]
Still nothing? 😔

**Comment by paclw-cory** [comment]
I had a case where I learned to trigger form validation errors from a course on Udemy.

This triggered the missing dependency "dispatch", but still worked:

```
  useEffect(() => {
    if (state.business_name.value) {
      const delay = setTimeout(() => dispatch({ type: "businessNameAfterDelay" }), 1000)
      return () => clearTimeout(delay)
    }
  }, [state.business_name.value])

```
Using @ra30r 's example, I changed it to this to clear the error:

```
  const businessNameAfterDelay = function () {
    if (state.name.value) {
      const delay = setTimeout(() => dispatch({ type: "businessNameAfterDelay" }), 1000)
      return () => clearTimeout(delay)
    }
  }
  useEffect(businessNameAfterDelay, [state.business_name.value])
```

**Comment by AprilRood** [comment]
I don't know if I have any idea what I'm talking about, but I got the impression that I was getting this error when I reference a variable in useEffect before I have "set" that variable. Here is my code when I was getting the error:

` const id = props.match.params.id;
  **const [roster, setRoster] = useState([]);**
  const [playerId, setPlayerId] = useState("");
  **const [players, setPlayers] = useState([]);**

  useEffect(() => {
    const fetchdata = async () => {
      Axios.get(process.env + "team-players/players/" + id)
        .then((res) => {
          res.data.map((player) => {
            Axios.get(process.env + player.player)
              .then((response) => {
                setPlayerId(response.data._id);
                **players.push(response.data);**
                **setRoster(players.concat(roster));**
              })
              .catch((err) => console.log(err));
          });
        })
        .catch((err) => console.log(err));
    };
    fetchdata();
  }, [id]);`

And I was getting this error: "_React Hook useEffect has missing dependencies: 'players' and 'roster'. Either include them or remove the dependency array._" As you can see, players is accessed with "players.push(response.data)" without using "setPlayers" as defined in the const above. Same with roster accessing itself with "setRoster(players.concat(roster))" without having set it with "setRoster".

So, I added:
`  ****setPlayers([]);**
   **setRoster([])**;**`

Just below useEffect and the error is gone.

`useEffect(() => {
    **setPlayers([]);
    setRoster([]);**
    const fetchdata = async () => {
      Axios.get(process.env + "team-players/players/" + id)
        .then((res) => {
          res.data.map((player) => {
            Axios.get(process.env + player.player)
              .then((response) => {
                setPlayerId(response.data._id);
                players.push(response.data);
                setRoster(players.concat(roster));
              })
              .catch((err) => console.log(err));
          });
        })
        .catch((err) => console.log(err));
    };
    fetchdata();

  }, [id]);`

I hope that helps someone.

**Comment by RichardRosario** [comment]
> Why is this closed?

**I believe an answer has been given somewhere.. lol.. anyway.. the solution is to do like this:**
`  useEffect(() => {
    dispatch();
  }, [dispatch]);`

hope this helps!!

**Comment by gugol2** [comment]
> > Why is this closed?
> 
> **I believe an answer has been given somewhere.. lol.. anyway.. the solution is to do like this:**
> ` useEffect(() => { dispatch(); }, [dispatch]);`
> 
> hope this helps!!

Yes, that workaround works, @kennylbj said here some time ago.
Why the issue is closed? There are memes here also about it... xd

**Comment by spmsupun** [comment]
My issue kinda like this (sample code):

```
  const [userStatus, setUserStatus]: any = React.useState([]);
  const [userData, setUserData]: any = React.useState([]);

  useEffect(() => {
    setUserStatus(!userStatus);
    return () => {};
  }, [userData]);
```
So as in the code you want to change `userStatus` value when there's a change in `userData` and in this case if you want to check `userStatus` value you have add it to deps:

```
  const [userStatus, setUserStatus]: any = React.useState([]);
  const [userData, setUserData]: any = React.useState([]);

  useEffect(() => {
    setUserStatus(!userStatus);
    return () => {};
  }, [userData, userStatus]);
```
in this scenario it will be a never ending loop

**Comment by kennylbj** [comment]
@spmsupun 
You can use the callback-style of useStatus to avoid introducing userStatus variable in the dependencies array of useEffect.

```ts
 useEffect(() => {
    setUserStatus(prevState => !prevState);
    return () => {};
  }, [userData]);
```

**Comment by AjayYadav7** [comment]
>Solution -->just add dispatch in last  for dependency
> **React Hook useEffect has a missing dependency: 'dispatch'.**
> 
> ```
>   useEffect(() => {
>     axios.get('http://localhost:8000/api/future-registrations')
>       .then((result) => {
>         dispatch(initRegistrationsAction(result.data));
>       });
>   }, [**dispatch**]);
> ```

**Comment by spmsupun** [comment]
@kennylbj This is good when you deal with one state, my actual scenario is with props. I am sending a props from parent and it change twice so usereffect calls twice but I am running a listener inside it and it will listen twice.

**Comment by kennylbj** [comment]
@spmsupun But I think it's quite normal to unregister the previous listener and register a new listener whenever the variables in dependencies array changed(eg. when the userId in props changed, the listener also needed to be changed). And this behavior is what react expects to have.

If you still don't want to make the listener called twice, I think you should take more care of the props passed to it and make sure it will not be changed during renders.

**Comment by spmsupun** [comment]
@kennylbj well I do unregister but still it not a good practise
![image](https://user-images.githubusercontent.com/1854023/88365903-1858eb80-cda5-11ea-8615-70749c3ae888.png)
Guess I have to handle this from listener class,

**Comment by kennylbj** [comment]
Seems like your listener is called twice even if the socket id are the same.

I don't think this will happen if the following situation meets:

 ```ts
// call useEffect if and only if sockeId changed.
useEffect(() => {
  const unregister = register(socketId);
  return () => unregister(socketId);
}, [socketId])
```

Are there any other variables in the dependencies array changed during different renders to cause this behavior?

**Comment by spmsupun** [comment]
Yeah 3 dependencies, but I figure it out some different way

**Comment by uc-asv** [comment]
Is anyone find the solution?

I am getting the warning at this code

```
useEffect(() => {
        axios.get('http://www.omdbapi.com/?apikey=32395055&type=movie&s=bad&y='+state.year_filter)
            .then(res => {
                dispatch({type: 'FETCH_SUCCESS',movieDetails: res.data.Search, response: res.data.Response, error: (res.data.Error !== undefined) ? res.data.Error: ''});
            })
            .catch(e => {
                dispatch({type: 'FETCH_ERROR', error: 'Something Went wrong', response: 'False'});
            })
    }, [state.year_filter])
```
Warning :
`React Hook useEffect has a missing dependency: 'dispatch'. Either include it or remove the dependency array  react-hooks/exhaustive-deps`

**Comment by shinoshu** [comment]
ぴえん

**Comment by weiwentao518** [comment]
> As suggested in the warning message, you can do like that
> 
> ```
> const initFetch = useCallback(() => {
>     dispatch(fetchPosts());
>   }, [dispatch]);
> 
>   useEffect(() => {
>     initFetch();
>   }, [initFetch]);
> ```

wow，非常感谢你🙏

**Comment by ahtashamabbasse** [comment]
🚀 Below code snippet works really well for me 

**On Component did mount**

```
const onInit = () => getUsers()
useEffect(onInit, [])
```

**On parameter change**

```
const onInit = () => getUser(id)
useEffect(onInit, [id])
```

**Comment by ymoon715** [comment]
Best work-around: react-hooks/exhaustive-deps: "off"
I have never found this eslint feature to be useful. Ever.

**Comment by taotao9125** [comment]
> 🚀 Below code snippet works really well for me
> 
> **On Component did mount**
> 
> ```
> const onInit = () => getUsers()
> useEffect(onInit, [])
> ```
> 
> **On parameter change**
> 
> ```
> const onInit = () => getUser(id)
> useEffect(onInit, [id])
> ```

nice work.

**Comment by k-funk** [comment]
I don't think the react docs quite cover @luojinghui 's use case, that is also the same that I have:

We have multiple ways of calling a given function, ie: `setCenterPosition()`. Controlled by state (hooked up by `useEffect`), click handler (called directly), etc.

The react docs suggest
> This is why **usually you’ll want to declare functions needed by an effect inside of it**

But usually in real-world examples, we need to reuse functions like we normally would with instance methods on a class component (`this.setCenterPosition(...)`), and I'm starting to wonder if hooks are actually good for feature-full, complex components.

I'm not actually sure, but maybe the solution is just to make a callback

```
const setCenterPosition = useCallback(() => {
  ...
}, [Stage, elePositionArr, setElePositionArr]);

useEffect() => {
  ...
}, [stageRef, Stage, currentIndex, setCenterPosition]);
```

...but that's kinda gross to me, and I'm hoping there's a cleaner way. Maybe @gaearon can confirm?

**Comment by naveenkash** [comment]
facing the same issue i can disable the warning bu using this line at the end of useEffect `
    // eslint-disable-next-line react-hooks/exhaustive-deps` but i don't want to use it

Here's My code

`const [items, setItems] = useState([
    {
      name: "test 1",
      id: 1
    },
    {
      name: "test 2",
      id: 2
    },
    {
      name: "test 3",
      id: 3
    },
    {
      name: "test 4",
      id: 4
    },
    {
      name: "test 5",
      id: 5
    }
  ]);

  useEffect(() => {
    const intervalId = setInterval(() => {
      setItems(shuffleArray(items));
    }, 1000);
    return () => {
      clearInterval(intervalId);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

**Comment by ewmiller** [comment]
My solution to this problem has been to simply stop using React Hooks. If you component needs lots of state logic and lifecycle events like `componentDidMount` that badly, just use a class component and save yourself the headache. Class components are perfectly fine, and if this thread proves anything, it's that React Hooks are not ready to fully replace them when complex state logic or lifecycle events are needed (nor do they confer any advantage - is your code _really_ that much better for having used React Hooks?)

**Comment by rommyarb** [comment]
This is my solution for now:
```jsx
const mounted = () => {
  dispatch(something());
}

useEffect(mounted, []);
```

Thank you @ra30r [react/issues/15865#issuecomment-651254164](https://github.com/facebook/react/issues/15865#issuecomment-651254164)

**Comment by ManelGonzalez-ops** [comment]
What's the point on adding the function into the dependency Array? What's the point on those warnings? Kinda stupid

**Comment by k-funk** [comment]
> What's the point on adding the function into the dependency Array? What's the point on those warnings? Kinda stupid

Please read an article on the topic. https://reacttraining.com/blog/when-to-use-functions-in-hooks-dependency-array/

**Comment by avillarubia** [comment]
this works for me

```
const dispatch = useDispatch()

    useEffect(() => {
        async function fetch() {
            try {
                const { data } = await getDiscussionByTitle(params.title)
                setDiscussion(data)
            } catch (error) {
                dispatch(addRequestError({ message: error.response.data }))
            }
        }

        fetch()
    }, [params.title, dispatch])

```

**Comment by kwabena53** [comment]
> As suggested in the warning message, you can do like that
> 
> ```
> const initFetch = useCallback(() => {
>     dispatch(fetchPosts());
>   }, [dispatch]);
> 
>   useEffect(() => {
>     initFetch();
>   }, [initFetch]);
> ```

This is what solved my problem. Don't forget to 
`import  React,` { useCallback } from "react"`

**Comment by GabeCabrera** [comment]
This is a reasonably common issue, which can be solved by passing your dependency into the dependency array.

In the OP's example it could be solved by passing `setCenterPosition` like this.
In any of your cases, you would pass your dependency in the same place, this is not specific to his question.

![carbon](https://user-images.githubusercontent.com/8227426/107451592-74dee680-6afc-11eb-80ef-3b76f27ffdb5.png)

For more information, as @gaearon said, it can be found here:
https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies

If this warning is not handled it can cause memory leaks in your application as it can cause infinite loops and re-render issues.

**Comment by vincent38wargnier** [comment]
Did anyone find a way to call a function inside a useEffect with empty array callback without adding bullshit code? 

React has the power to make easy things impossible. I love it.

**Comment by Kaydhiman** [comment]
> > This seems to be quite a common use case - one that I have run into on my current project.
> > You want to run an effect hook just one time, but there IS a dependancy, although you only care about the state of that at startup. Currently using
> > `// eslint-disable-next-line react-hooks/exhaustive-deps`,
> > to turn off the linting rule but would be nice to not have to in this case.
> 
> Closing Lint is just a temporary solution, the key to this problem is not enough to understand the hook.

I agree with you @luojinghui Please reopen this issue. I am want to know the perfect solution for same...

**Comment by AKclown** [comment]
I want to execute useEffect only when specific attributes are changed, not the dependent parameters in useEffect. But eslint will warn me, how to solve it

**Comment by domotisas** [comment]
Hi, a solution is to simule class constructor like this: https://dev.to/bytebodger/constructors-in-functional-components-with-hooks-280m

**Comment by RacketyWater7** [comment]
> I want to execute useEffect only when specific attributes are changed, not the dependent parameters in useEffect. But eslint will warn me, how to solve it

Well then put those attributes into the dependency array, and as for avoiding the eslint warning: also put the eslint line (below added) at the end of the useEffect to not let it call warning and work as expected hopefully.
// eslint-disable-next-line react-hooks/exhaustive-deps

**Comment by yoieh** [comment]
This still feels like an issue... Or am i just dumb?

Example:
```js
const [id, setId] = useState(null)
const [stateData, setStateData] = useState()

useEffect(() => {
  const fetchData = async () => {
      const data = fetch(`/data/${id}`);
  
      if (data) {
          setData(data);
      }
  };
  
  if (id) { // only fetch data is id is set
      fetchData();
  }
}, [id]); // React Hook useEffect has a missing dependency: 'setData'. Either include it or remove the dependency array.
```


Example with out warrning
```js
const [id, setId] = useState(null)
const [stateData, setStateData] = useState()

useEffect(() => {
  const fetchData = async () => {
      const data = fetch(`/data/${id}`);
  
      if (data) {
          setData(data);
      }
  };
  
  if (id) { // only fetch data is id is set
      fetchData();
  }
}, [id, setData]); // No warning but results in a infinity loop
```

**Comment by k-funk** [comment]
> This still feels like an issue or at least a bit confusing... Or am i just dumb and misusing my useEffect?
> 
> Example:
> 
> ```js
> const [id, setId] = useState(null)
> const [stateData, setStateData] = useState()
> 
> useEffect(() => {
>   const fetchData = async () => {
>       const data = fetch(`/data/${id}`);
>   
>       if (data) {
>           setData(data);
>       }
>   };
>   
>   if (id) { // only fetch data if id is set
>       fetchData();
>   }
> }, [id]); // React Hook useEffect has a missing dependency: 'setData'. Either include it or remove the dependency array.
> ```
> 
> Example with out warrning
> 
> ```js
> const [id, setId] = useState(null)
> const [stateData, setStateData] = useState()
> 
> useEffect(() => {
>   const fetchData = async () => {
>       const data = fetch(`/data/${id}`);
>   
>       if (data) {
>           setData(data);
>       }
>   };
>   
>   if (id) { // only fetch data if id is set
>       fetchData();
>   }
> }, [id, setData]); // No warning but results in a infinity loop
> ```

I'm presuming that your variable names are mixed up, and that `setData()` is actually `setStateData()`. If so, the reason that you don't need to put `useState` setters in the dependency is array is because

> React guarantees that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list

Source: https://reactjs.org/docs/hooks-reference.html#usestate

However, it should not be rerunning the `useEffect` infinitely. I would debug some more and make sure your variable names are what you expect them to be.

If by chance your `setData` is actually an unstable function (passed down in props?), that could cause an infinite loop.

**Comment by yoieh** [comment]
> I'm presuming that your variable names are mixed up, and that `setData()` is actually `setStateData()`. If so, the reason that you don't need to put `useState` setters in the dependency is array is because

you are right, it should be `setStateData()`

And yes its an function exported passed down by a context

```js
const [stateData, setStateData] = useState()

const setStateData = (key: string, value: any) => {
    setState((prev) => {
        return {
            ...prev,
            [key]: value,
        };
    });
};

....

return (
      <StateContext.Provider
          value={{
              ...state,
            setStateData
          }}
      >
          {children}
      </StateContext.Provider>
  );
};

export const useStateContext = () => React.useContext(StateContext);
```

Then used like this.
```js
const [id, setId] = useState(null)
const {stateData, setStateData} = useStateContext()

useEffect(() => {
  const fetchData = async () => {
      const data = fetch(`/data/${id}`);
  
      if (data) {
          setData(data);
      }
  };
  
  if (id) { // only fetch data if id is set
      fetchData();
  }
}, [id, setData]); // No warning but results in a infinity loop
```

Sow it seems to be me being dumb and using it wrong.

**Comment by k-funk** [comment]
> > I'm presuming that your variable names are mixed up, and that `setData()` is actually `setStateData()`. If so, the reason that you don't need to put `useState` setters in the dependency is array is because
> 
> you are right, it should be `setStateData()`
> 
> > If by chance your setData is actually an unstable function (passed down in props?), that could cause an infinite loop.
> 
> And yes it's an function exported and passed down by a context..
> 
> ```js
> const [stateData, setStateData] = useState()
> 
> const setStateData = (key: string, value: any) => {
>     setState((prev) => {
>         return {
>             ...prev,
>             [key]: value,
>         };
>     });
> };
> 
> ....
> 
> return (
>       <StateContext.Provider
>           value={{
>               ...state,
>             setStateData
>           }}
>       >
>           {children}
>       </StateContext.Provider>
>   );
> };
> 
> export const useStateContext = () => React.useContext(StateContext);
> ```
> 
> Then used like this.
> 
> ```js
> const [id, setId] = useState(null)
> const {stateData, setStateData} = useStateContext()
> 
> useEffect(() => {
>   const fetchData = async () => {
>       const data = fetch(`/data/${id}`);
>   
>       if (data) {
>           setData(data);
>       }
>   };
>   
>   if (id) { // only fetch data if id is set
>       fetchData();
>   }
> }, [id, setData]); // No warning but results in a infinity loop
> ```
> 
> So it seems to be me being dumb and using it wrong.

Not dumb. I also find hook dependencies to be rather confusing, and often screw up my code without realizing it.

In your case, you can probably use `useCallback()` to make your function stable:
```
const setStateData = (key: string, value: any) => {
  setState(...)
}
```

-> 

```
const setStateData = useCallback((key: string, value: any) => {
  setState(...)
}, []) // shouldn't need `setState` as a dep
```

**Comment by Fiaz-Tariq** [comment]
To fix all hook warnings you can add the below code in the package.json file below the eslintConfig:
{
  "plugins": [
    "react-hooks"
  ],
  "rules": {
    "react-hooks/rules-of-hooks": "error",
    "react-hooks/exhaustive-deps": "off"
  }
}
